ibis/backends/base/sqlglot/compiler.py

```diff
@@ -0,0 +1,1300 @@
+from ibis.common.patterns import replace
+from ibis.expr.analysis import p, x
+from ibis.expr.rewrites import replace_scalar_parameter, unwrap_scalar_parameter
+
+if TYPE_CHECKING:
+    import ibis.expr.schema as sch
+    import ibis.expr.types as ir
+    from ibis.backends.base.sqlglot.datatypes import SqlglotType
+
+
+class AggGen:
+    __slots__ = ("aggfunc",)
+
+    def __init__(self, *, aggfunc: Callable) -> None:
+        self.aggfunc = aggfunc
+
+    def __getattr__(self, name: str) -> partial:
+        return partial(self.aggfunc, name)
+
+    def __getitem__(self, key: str) -> partial:
+        return getattr(self, key)
+
+
+class FuncGen:
+    __slots__ = ()
+
+    def __getattr__(self, name: str) -> partial:
+        return lambda *args, **kwargs: sg.func(name, *map(sge.convert, args), **kwargs)
+
+    def __getitem__(self, key: str) -> partial:
+        return getattr(self, key)
+
+    def array(self, *args):
+        return sge.Array.from_arg_list(list(map(sge.convert, args)))
+
+    def tuple(self, *args):
+        return sg.func("tuple", *map(sge.convert, args))
+
+    def exists(self, query):
+        return sge.Exists(this=query)
+
+    def concat(self, *args):
+        return sge.Concat(expressions=list(map(sge.convert, args)))
+
+    def map(self, keys, values):
+        return sge.Map(keys=keys, values=values)
+
+
+class ColGen:
+    __slots__ = ()
+
+    def __getattr__(self, name: str) -> sge.Column:
+        return sg.column(name)
+
+    def __getitem__(self, key: str) -> sge.Column:
+        return sg.column(key)
+
+
+def paren(expr):
+    """Wrap a sqlglot expression in parentheses."""
+    return sge.Paren(this=expr)
+
+
+def parenthesize(op, arg):
+    if isinstance(op, (ops.Binary, ops.Unary)):
+        return paren(arg)
+    # function calls don't need parens
+    return arg
+
+
+C = ColGen()
+F = FuncGen()
+NULL = sge.NULL
+FALSE = sge.FALSE
+TRUE = sge.TRUE
+STAR = sge.Star()
+
+
+@replace(p.InValues(..., ()))
+def empty_in_values_right_side(_):
+    """Replace checks against an empty right side with `False`."""
+    return ops.Literal(False, dtype=dt.bool)
+
+
+@replace(
+    p.WindowFunction(
+        p.PercentRank(x) | p.RankBase(x) | p.CumeDist(x) | p.NTile(x),
+        p.WindowFrame(..., order_by=()) >> _.copy(order_by=(x,)),
+    )
+)
+def add_order_by_to_empty_ranking_window_functions(_):
+    """Add an ORDER BY clause to rank window functions that don't have one."""
+    return _
+
+
+@replace(
+    p.WindowFunction(p.RankBase | p.NTile)
+    | p.StringFind
+    | p.FindInSet
+    | p.ArrayPosition
+)
+def one_to_zero_index(_, **__):
+    """Subtract one from one-index functions."""
+    return ops.Subtract(_, 1)
+
+
+@replace(ops.NthValue)
+def add_one_to_nth_value_input(_, **__):
+    if isinstance(_.nth, ops.Literal):
+        nth = ops.Literal(_.nth.value + 1, dtype=_.nth.dtype)
+    else:
+        nth = ops.Add(_.nth, 1)
+    return _.copy(nth=nth)
+
+
+@public
+class SQLGlotCompiler(abc.ABC):
```

Comment:
Should this be `Concrete` instead? Something like:

cass SQLGlotCompiler(Concrete):

    @attribute
    def f(self):
        return FuncGen()

    @attribute
    def agg(self):
        return AggGen()
```

Reply:
Just took a look at the base compiler and the clickhouse one and both seem to be stateless. Could we convert the compiler to be just a collection of classmethods instead?

Reply:
Perhaps, but I'd rather do it in a follow up because that is a lot of churn for the current code.

Reply:
SGTM